### ISSUE TITLE
Make full text search results "scrollable"

### ISSUE THREAD

**Issue by marcderbauer** [marcderbauer/songcrawler#37]
Currrently only displays the first 10 search results when doing a full-text search.  
I thought it would be enough, but as it turns out the spotify search can be a bit difficult sometimes.  
  
I tried `python3 main "artist:Drake"` which should be pretty easy as Drake is one of the biggest artists around currently, but he doesn't show up in the first 10 search results?  I guess the name is quite generic, but still not great.  

Already created a dummy method `_user_select_spotify()` which is supposed to do the following:
  
        Prints the first 10 search results. The user can select one of the items or type get the next batch.
        ">/n/next" for next, "</p/previous" for previous
        [0-9] for selection

It's important to get the part where the API query is made into the the search loop.  

It's probably going to be something like this:
  
    offset = 0
    while(True):

        query API(offset)
        pretty_print results
        selection = user_select results -> returns uri or "previous" or "next"
        
        if "previous":
                offset -=10 (make sure it doesn't go past 0 !!!)
        elif "next": 
                offset += 10
        else:
                return selection

**Comment by marcderbauer** [comment]
After issue #35, things are now split into model and controller, with the view being baked into the model. It would make sense to use the MVC principle and separate the view completely. Might use the `npyscreen` library for that. Ideally it would be an actually scrollable, full-screen view.  
  
Here's an introductory medium article on it:
https://medium.com/@ValTron/create-tui-on-python-71377849879d

**Comment by marcderbauer** [comment]
Turns out npyscreen hasn't been updated to work on python > 3.7. 
There's a `RuntimeError: generator raised StopIteration` error. It may be possible by installing the library locally and digging into the source code, but this would be hard to make reproducable / to share on GitHub.  
  
As the main focus is on developing something easy to use right away I will look into other libraries

**Comment by marcderbauer** [comment]
Decided to use [PyTermGui](https://ptg.bczsalba.com/).  
  
Rough Idea:
One big surrounding window
One container for each retrieved result, maybe formatted something like this:
  
    -------------------------------------------
    Song: Till Death
    Album: Soft Sounds From Another Planet
    Artist: Japanese Breakfast
    -------------------------------------------
Width could be something like 70% for this
Could keep the header and footer with quit button from the example app

**Comment by marcderbauer** [comment]
Got a MVP to work now (to a certain degree).
Cool thing is: It's actually scrollable with the mousewheel. But there's two issues:
1. It's too big / there's too much space everywhere.
2. I don't know if / how you can scroll with keys
<img width="939" alt="Screenshot 2022-11-26 at 12 24 46" src="https://user-images.githubusercontent.com/29900658/204086397-a579fa1e-35c0-4b67-99e9-de0acc85378d.png">

I'd rather revert back to the old layout (but still using pytermgui):

| index | Track              | Album                | Artist          |
|-------|--------------------|----------------------|-----------------|
| 0     | Love Story         | Cavalcovers          | Black Midi      |
| 1     | A Tooth for an Eye | Shaking The Habitual | The Knife       |
| 2     | Wolf               | Cool It Down         | Yeah Yeah Yeahs |

Basically what I already have, just with pytermgui
<img width="824" alt="image" src="https://user-images.githubusercontent.com/29900658/204086545-7eabbe82-9968-4a97-bae4-c843b9013092.png">


## How to do this:
1. A general header component
2. One component / widget which defines the rows

The rows (as well as the header) should be a container which contains multiple subcontainers which are split and placed next to one another.
I need the same funcitonality for the header. Might aswell make them the same class and if I need to I can create subclasses based on that.

**Comment by marcderbauer** [comment]
Managed to get the formatting a into rows and columns. A lot better already.
Started a new Class `Table`, which handles formatting the data into a Table.
  
Currently it has a `list_to_row()` method, which does exactly what the name might imply.
Next up is adding in funcionality for a header, and some way to turn lists of lists into lists of rows.
This, together with the header can then be used to initialise a table in one go (eventually)

<img width="1074" alt="image" src="https://user-images.githubusercontent.com/29900658/204092732-a4ffcaf5-ca98-4ccc-be82-7aa71a85a32d.png">

**Comment by marcderbauer** [comment]
Improved the `Table` class, with `append_row()` and append_rows()` taking both lists and containters.
Header is implemented as well now
<img width="1051" alt="image" src="https://user-images.githubusercontent.com/29900658/204094600-e300f5e1-6ef7-42eb-9642-4ff1cae49f04.png">
  
The next important thing is probably to make the index smaller. I don't like how much space it takes up. Will make the rest of the text look better too.

**Comment by marcderbauer** [comment]
For reference: This is what `box="EMPTY_VERTICAL" looks like. The horizontal lines don't align and everything takes up a lot of space.
<img width="1051" alt="image" src="https://user-images.githubusercontent.com/29900658/204094772-0cd54918-1d79-411e-ab75-bf74329f36b4.png">

**Comment by marcderbauer** [comment]
It looks a lot better when it's left aligned. 

<img width="928" alt="Screenshot 2022-11-26 at 17 41 46" src="https://user-images.githubusercontent.com/29900658/204099482-967e8ffe-655d-4745-85d5-21aa30050a4c.png">


Tried seperating the index row from the other rows and then wrapping another splitter around the combined row, but that wouldn't change the index width unfortunately...

Will try around a bit more, but might just create a GitHub issue

**Comment by marcderbauer** [comment]
Could probably make each line a button which could be pressed? Index would just be a part of the string then. 

Look at example here:
https://ptg-docs-beta.netlify.app/widgets/styling/

**Comment by marcderbauer** [comment]
Just going to go with the wide index for now. Might switch to `https://github.com/Textualize/textual` at some point in the future, but I want to move on. If the view is implemented correctly it should be independent from the rest of the implementation anyway.  

Just created a `View` class, which should be used by `Songcrawler` to display search results.  

## How it should work:
The method in music should not just be `search`, but something like `search_spotify` or `get_search_results`. The actual search should move up into the `Songcrawler` class. 
`search_spotify` should just return the `search_results`, which it already gets. Might make sense to turn it into a list of dicts:
  
    [{songname: x, albumname:y, artist_name: z, uri=???},{...},...]
  
Like this you could easily include the uri in the search results. However, the view class only takes lists, so maybe not that great of an idea?
Anyway, you instantiate the view class, which instantiates a table (view.table). To this you can append the rows and header. Might want to create a method to do all of this in one go, maybe through a class method.
You start the view and select one of the results.  
  
The view will return the index you selected and you can use that to search the uri which can then be passed onto the rest of the program

**Comment by marcderbauer** [comment]
I need to take into account that the search can only be used as a command-line interface, not when importing Songcrawler as a class.

**Comment by marcderbauer** [comment]
Solution for when importing (non-CLI):
Add an additional parameter `interactive=False` to the search function (or `request` or whatever it's called).  
`main.py` then only calls the function with `interactive=True` as it provides the CLI functionality, but users importing songcrawler will not have to deal with that.

**Comment by marcderbauer** [comment]
The whole library has been a bit of a pain to use. Tried using textual but found it to be overkill for just printing a table...
Going to just use [Rich](https://github.com/Textualize/rich) now.  
  
Focus is on getting the `View` class out. The rest I can deal with then.

**Comment by marcderbauer** [comment]
I just finished writing a first version of the view class. It's integrated into Songcrawler as well now and works great!
<img width="911" alt="Screenshot 2022-11-28 at 13 03 25" src="https://user-images.githubusercontent.com/29900658/204273253-a2fbcda0-50f2-45aa-b53a-e5a62136c027.png">
  
Also added the `interactive` attribute to Songcrawler.
If it is disabled (calling through import, not CLI), then it will return the following dataclass:  
  
    class SearchResult:
    header:list[str]
    rows:list[list[str]]
    uris:list

Will need to clean up the code a bit now, I can remove all the other functions and files I created for the previous views. Otherwise will have to look through my Todos here.

Still got the Problem with Drake not working. For whatever reason he won't even show up in the first 50 results when searching for artists. Seems like it's an issue on Spotify's end though, as all other queries work fine and you still get 50 other results.  
  
**TODO:**
The last big thing for the view is adding a previous/next feature. Returning 50 results is too much and you likely will never need that many. I just changed the limit to `15` as it's a good number, which will likely have all results, but it doesn't take up the entire screen.  
  
I might just query the first 50 results and then use prev/next to iterate through those rather than re-requesting? Will need to see how difficult that is.
